File: pbmx_ffi/java/io/rmf/pbmx/Id.java
package io.rmf.pbmx;

import static io.rmf.pbmx.Util.toMaskArray;
import io.rmf.pbmx.ffi.LibPbmx;
import io.rmf.pbmx.ffi.RawMask;
import io.rmf.pbmx.ffi.RawFingerprint;
import java.util.Collection;

public final class Id {
    RawFingerprint raw;

    Id(RawFingerprint raw) {
        this.raw = raw;
    }

    public static Id of(Collection<Mask> stack) {
        RawMask[] masks = toMaskArray(stack);
        return new Id(LibPbmx.INSTANCE.pbmx_stack_id(masks, masks.length));
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        for (byte b : this.raw.bytes) {
            sb.append(String.format("%02x", b));
        }
        return sb.toString();
    }
}

File: pbmx_droid/app/src/main/java/io/rmf/pbmx/Id.kt
package io.rmf.pbmx

import io.rmf.pbmx.ffi.FFI
import io.rmf.pbmx.ffi.RawFingerprint

class Id internal constructor(internal var raw: RawFingerprint): Comparable<Id> {

    override fun equals(other: Any?): Boolean {
        return this.toString() == other.toString()
    }

    override fun hashCode(): Int {
        return this.toString().hashCode()
    }

    override fun compareTo(other: Id): Int {
        return this.toString().compareTo(other.toString())
    }

    override fun toString(): String {
        return this.raw.bytes
            .fold(StringBuilder()) { sb, b -> sb.append(String.format("%02x", b)) }
            .toString()
    }

    companion object {

        fun of(stack: Collection<Mask>): Id {
            val masks = toMaskArray(stack)
            return Id(FFI.pbmx_stack_id(masks, masks.size.toLong()))
        }
    }
}

